File: examples/basic/annotations/size_bar.py
"""
Example demonstrating the SizeBar annotation for radial glyphs.

SizeBar provides a visual reference for understanding the size mapping
in scatter plots where circle radius varies with data values.
"""

import numpy as np

from bokeh.models import SizeBar
from bokeh.plotting import figure, show

N = 100
x = np.random.random(size=N) * 100
y = np.random.random(size=N) * 100
radii = np.random.random(size=N) * 10
colors = np.array([(r, g, 150) for r, g in zip(50 + 2*x, 30 + 2*y)], dtype=np.uint8)

p = figure()
cr = p.circle(x, y, radius=radii, fill_color=colors, fill_alpha=0.6, line_color=None)

size_bar = SizeBar(
    renderer=cr,
    title="SizeBar component",
    width="max",
    orientation="horizontal",
    glyph_fill_color="violet", glyph_fill_alpha=0.8, glyph_line_color="black",
    border_line_color="gray", border_line_dash="dotted",
)
p.add_layout(size_bar, "below")

show(p)

File: bokehjs/test/integration/annotations/size_bar.ts
import {display, fig} from "../_util"

import {SizeBar} from "@bokehjs/models"
import type {Side, Orientation} from "@bokehjs/core/enums"
import type {Size} from "@bokehjs/core/types"
import {Random} from "@bokehjs/core/util/random"

describe("SizeBar annotation", () => {
  function make_plot(size: Size = {width: 300, height: 200}) {
    const random = new Random(1)

    const N = 100
    const x = random.floats(N).map((v) => 100*v)
    const y = random.floats(N).map((v) => 100*v)
    const radii = random.floats(N).map((v) => 10*v)

    const p = fig([size.width, size.height])
    const cr = p.circle({x, y, radius: radii, line_color: "black", fill_alpha: 0.8})

    return {p, cr}
  }

  const common: Partial<SizeBar.Attrs> = {
    title: "SizeBar",
    glyph_fill_color: "pink",
    glyph_line_color: "green",
    border_line_color: "violet",
    border_line_dash: "dotdash",
    bar_line_color: "gray",
    bar_line_dash: "dotted",
  }

  it("should support automatic renderer", async () => {
    const {p} = make_plot()

    const size_bar = new SizeBar({renderer: "auto", orientation: "horizontal", width: "max", ...common})
    p.add_layout(size_bar, "below")

    await display(p)
  })

  it("should support manual renderer", async () => {
    const {p, cr} = make_plot()

    const size_bar = new SizeBar({renderer: cr, orientation: "horizontal", width: "max", ...common})
    p.add_layout(size_bar, "below")

    await display(p)
  })

  async function test(orientation: Orientation, side: Side, width: number, height: number) {
    const {p, cr} = make_plot({width, height})

    const size_bar = new SizeBar({renderer: cr, orientation, ...common})
    p.add_layout(size_bar, side)

    await display(p)
  }

  describe("should support horizontal orientation", () => {
    it("above", async () => test("horizontal", "above", 300, 200))
    it("below", async () => test("horizontal", "below", 300, 200))
    it("left",  async () => test("horizontal", "left",  500, 200))
    it("right", async () => test("horizontal", "right", 500, 200))
  })

  describe("should support vertical orientation", () => {
    it("above", async () => test("vertical", "above", 300, 400))
    it("below", async () => test("vertical", "below", 300, 400))
    it("left",  async () => test("vertical", "left",  300, 200))
    it("right", async () => test("vertical", "right", 300, 200))
  })
})
